File: saveWatcher.py
import sys,os, time,hashlib
from hashlib import md5

initalListOfFiles = []
filesToWatch = dict()
userRequestedFile = ""
targetFileExtension = ".sav"
secondsBetweenRefreshes = 5
archiveFolderName = "backupPilots"

#simplify file name retrieval
def files(path):
    for file in os.listdir(path):
        if os.path.isfile(os.path.join(path, file)):
            yield file

#get the initial list of files
for file in files("."):
    initalListOfFiles.append(file)

# All files, or just a specific file?
if(len(userRequestedFile)<=0):
    print("No specific file to watch was requested.")
    print("Watching all *"+targetFileExtension+" files in this folder.")
    for file in initalListOfFiles:
        if (targetFileExtension in file):
            filesToWatch[file] = ''
else:
    for file in initalListOfFiles:
        if(file == userRequestedFile):
            filesToWatch.append(file)

print(" ")

#bailout if nothing to watch
if(len(filesToWatch) <= 0):
    print("Found no files to watch.")
    sys.exit(0)


#seed proper file structure to work with
filesAndHashes = {}
print("Found "+str(len(filesToWatch))+" file(s) to watch:")
for filename in filesToWatch:
    print("Watching "+filename)

print(" ")
print("Use CTRL-C to exit this watcher ")
print(" ")

#create the archive folder if we need to
archiveFolderPath = os.path.join(".",archiveFolderName)
if not os.path.exists(archiveFolderPath):
    os.mkdir(archiveFolderPath)

sys.stdout.flush() #had to flush to make sure the console stays updated before it starts spinning on the files

#main thread
try:
    while True:
        for filename in filesToWatch:
            #create a blank hash if the hash list hasn't seen the file yet
            if(filename not in filesAndHashes):
                filesAndHashes[filename] = ""

            try:
                #open and read in the file
                file = open(filename, 'rb')
                filecontent = file.read()
                file.close();

                #hash the file contents
                h = md5()
                h.update(str(filecontent).encode("utf-8"))

                #if the latest hash doesn't match the last hash for the file
                #then archive the file with a timestamp in the file name
                if(h.hexdigest() != filesAndHashes[filename]):
                    filesAndHashes[filename] = h.hexdigest()

                    timestamp = time.localtime();
                    print("updated " + filename + " was backed up at " + str(time.strftime("%H:%M:%S",timestamp)))

                    newFileName = str(time.strftime("%Y%m%d_%H%M%S",timestamp)) + "-" + filename
                    newFilePath = os.path.join(archiveFolderPath,newFileName);
                    newFile = open(newFilePath, mode="wb")
                    newFile.write(filecontent)
                    newFile.close()
            #IO Errors should throw a soft warning but aren't cause for alarm; we may get it backed up next pass
            except IOError  as e:
                print("burped on "+filename+" with taste: "+str(e))

            sys.stdout.flush()
            time.sleep(secondsBetweenRefreshes);

except(KeyboardInterrupt): #CTRL-C to exit
    sys.exit(0);

File: saveWatcher.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"time"
)

var initalListOfFiles []os.FileInfo

var userRequestedFile string = ""
var targetFileExtension string = ".txt"
var timeBetweenRefreshes string = "3s"
var archiveFolderName string = "backupPilots"

func main() {
    fmt.Println("This go program is incomplete and should not be expected to work.")
    fmt.Println("The python program in the same repository should work properly.")
    fmt.Println(" ")


	var secondsBetweenRefreshes time.Duration
	secondsBetweenRefreshes, err := time.ParseDuration(timeBetweenRefreshes)
	if err != nil {
		panic(err)
	}

	//get the inital list of files
	initalListOfFiles, err := ioutil.ReadDir("./")
	if err != nil {
		panic(err)
	}

	//set up the target list of files,
	//as only a handful have the right file extension
	// filename -> MD5sum
	var targetFilenameWithMD5sum map[string]string
	targetFilenameWithMD5sum = make(map[string]string)

	if len(userRequestedFile) <= 0 {
		fmt.Println("No specific file to watch was requested.")
		fmt.Println("Watching all *" + targetFileExtension + " files in this folder.")

		for _, file := range initalListOfFiles {
			if filepath.Ext(file.Name()) == targetFileExtension {
				targetFilenameWithMD5sum[file.Name()] = "<blank>"
			}
		}
	} else {
		fmt.Println("Looking for " + userRequestedFile + " as requested")
		for _, file := range initalListOfFiles {
			if file.Name() == userRequestedFile {
				targetFilenameWithMD5sum[file.Name()] = "<blank>"
			}
		}
	}

	fmt.Println(" ")

	if len(targetFilenameWithMD5sum) <= 0 {
		fmt.Println("Found no files to watch, exiting...")
		os.Exit(0)
	}

	fmt.Println("Found " + strconv.Itoa(len(targetFilenameWithMD5sum)) + " file(s) to watch.")
	for key := range targetFilenameWithMD5sum {
		fmt.Println("Watching:", key)
	}

	fmt.Println(" ")
	fmt.Println("Use CTRL-C to exit this watcher ")
	fmt.Println(" ")

	var done bool = false
	for !done {
		for key, value := range targetFilenameWithMD5sum {
			hashMD5, err := wrappedMD5sum(key)
			if err == nil {
				if(value != hashMD5){
                    targetFilenameWithMD5sum[key] = hashMD5
                    fmt.Println("updated "+key+" was backed up at " + value)
                }
			} else {
				fmt.Println("burped on " + key)
			}

		}
		time.Sleep(secondsBetweenRefreshes)
	}

}



func wrappedMD5sum(filename string) (string, error) {
	var md5ret string
	f, err := os.Open(filename)
	if err != nil {
		return md5ret, err
	}
	defer f.Close()

	h := md5.New()
	if _, err := io.Copy(h, f); err != nil {
		return md5ret, err
	}
	hashInBytes := h.Sum(nil)[:16]
	md5ret = hex.EncodeToString(hashInBytes)
	return md5ret, nil
}
